docs/topics/coroutines-basics.md

```diff
@@ -1,293 +1,536 @@
-allows `Done` string to be printed only after that:
+The `kotlinx.coroutines` library includes different dispatchers for different use cases.
+For example, [`Dispatchers.Default`](https://kotlinlang.org/api/kotlinx.coroutines/kotlinx-coroutines-core/kotlinx.coroutines/-dispatchers/-default.html) runs coroutines on a shared pool of threads performing work in the background,
+separate from the main thread.
 
-```text
-Hello
-World 1
-World 2
-Done
-```
+To specify a dispatcher for a coroutine builder like `.launch()`, pass it as an argument:
 
-<!--- TEST -->
+```kotlin
+suspend fun runWithDispatcher() = coroutineScope {
+    launch(Dispatchers.Default) {
+        println("Running on ${Thread.currentThread().name}")
+    }
+}
+```
 
-## An explicit job
+Alternatively, you can wrap multiple coroutines in a `withContext()` block to apply a dispatcher to all of them.
 
-A [launch] coroutine builder returns a [Job] object that is a handle to the launched coroutine and can be 
-used to wait for its completion explicitly.
-For example, you can wait for the completion of the child coroutine and then print the "Done" string:
+The following example runs on `Dispatchers.Default`, which is optimized for CPU-intensive operations like data processing:
 
 ```kotlin
+// Imports the kotlin.time.Duration to enable expressing duration in milliseconds
+import kotlin.time.Duration.Companion.milliseconds
+
 import kotlinx.coroutines.*
 
-fun main() = runBlocking {
-//sampleStart
-    val job = launch { // launch a new coroutine and keep a reference to its Job
-        delay(1000L)
-        println("World!")
-    }
-    println("Hello")
-    job.join() // wait until child coroutine completes
-    println("Done") 
-//sampleEnd    
+suspend fun main() = withContext(Dispatchers.Default) {
+   println("Running withContext block on ${Thread.currentThread().name}")
+
+   val one = async {
+      println("First calculation starting on ${Thread.currentThread().name}")
+      val sum = (1..500_000).sum()
+      delay(200L)
+      println("First calculation done on ${Thread.currentThread().name}")
+      sum
+   }
+
+   val two = async {
+      println("Second calculation starting on ${Thread.currentThread().name}")
+      val sum = (500_001..1_000_000).sum()
+      println("Second calculation done on ${Thread.currentThread().name}")
+      sum
+   }
+
+   // Waits for both calculations and prints the result
+   println("Combined total: ${one.await() + two.await()}")
 }
 ```
-{kotlin-runnable="true" kotlin-min-compiler-version="1.3"}
-<!--- KNIT example-basic-05.kt -->
-> You can get the full code [here](https://github.com/Kotlin/kotlinx.coroutines/blob/master/kotlinx-coroutines-core/jvm/test/guide/example-basic-05.kt).
->
-{style="note"}
+{kotlin-runnable="true"}
 
-This code produces: 
+> Even though coroutines can suspend and resume on different threads,
+> values written before the coroutine suspends are still available within the same coroutine when it resumes.
```

Comment:
```suggestion
> values written before the coroutine suspends are still guaranteed to be available within the same coroutine when it resumes.
```
I was reading this "still" as "even at that point in time" instead of "nonetheless" and got stumped.